File: Leetcode/cpp/Find Minimum in Rotated Sorted Array.cc
class Solution {
public:
    int findMin(vector<int> &num) {
        int n = num.size();
        return do_findMin(num, 0, n - 1);
    }
private:
    int do_findMin(vector<int>& num, int st, int end) {
        int l = st, r = end;
        if (l == r) {
            return num[l];
        }
        int res = min(num[st], num[end]);
        int mid = (l + r) >> 1;
        if (num[l] > num[mid]) {
            res = min(res, do_findMin(num, st, mid));
        } else {
            res = min(res, do_findMin(num, mid + 1, end));
        }
        return res;
    }
};

File: Leetcode/Algorithm/python/Find Minimum in Rotated Sorted Array.py
class Solution(object):
    def findMin(self, nums):
        n = len(nums)
        if n < 3:
            return min(nums)
        l, r = 0, n - 1
        while l <= r:
            m = (l + r) >> 1
            a, b, c = nums[l], nums[m], nums[r]
            if b > a > c:
                l = m
            elif b < c < a:
                r = m
            else:
                return min(a, b, c)
        return l
